src/csharp/Grpc.Core/Internal/ClientResponseStream.cs

```diff
@@ -49,19 +49,19 @@ public TResponse Current
 
         public async Task<bool> MoveNext(CancellationToken token)
         {
-            if (token != CancellationToken.None)
+            var cancellationTokenRegistration = token.CanBeCanceled ? token.Register(() => call.Cancel()) : (IDisposable) null;
+            using (cancellationTokenRegistration)
```

Comment:
There might be a timing dependent issue here:
- I need to unregister the cancellation token registration at some point and doing that once the read has finish seems to make sense
- the user expects the entire call to get cancelled whenever once he cancels the cancellation token, but there's effectively a race condition between the read finishing (and call.Cancel() invocation get unregistered) and user requesting the cancellation.

We actually might need to keep the cancellationTokenRegistration around until the call finishes.

Reply:
I think it's reasonable that if the cancellation comes after the read has completed, the cancellation token for just that read is irrelevant. Worth documenting, but then I think it's fine.

Reply:
In fact, more than fine: desirable. I could easily pass a new cancellation token into each MoveNext invocation, which auto-cancels after a timeout. I wouldn't want an "old" token to cancel the whole call when we've already used the result we created it for!

Reply:
Sg, leaving as is (+ added clarification to the docs).